FIX: Refactoring the ``B0FieldTransform`` implementation

oesteban opened the pull request:
This PR defines a new interface for the ``ApplyCoeffsField`` node, and dismisses ``TransformCoefficients`` because that projection is really not necessary.

Resolves: #345.

oesteban committed:
fix: refactoring the ``B0FieldTransform`` implementation

Addresses the issue of properly applying fieldmaps on distorted data.

Resolves: #345.

oesteban committed:
enh: draft the approx branch

effigies commented on sdcflows/transform.py:
```diff
@@ -79,47 +79,63 @@ def fit(self, target_reference, affine=None, approx=True):
+        # Project the reference's grid onto the fieldmap's
+        target_reference = target_reference.__class__(
+            target_reference.dataobj,
+            affine @ target_affine.T,
```
Is this right? Since `target_affine` is `IJK -> RAS(target)` and `affine` is `RAS(target) -> RAS(fieldmap)`, I think you want

```suggestion
            affine @ target_affine,
```

oesteban replied:
I believe you are right, but I'm not 100% sure so I opted for adding the `.T` and come back to it later. It really gives me confidence you noticed :).

So why I'm not 100% sure - the affine matrix has 4 column vectors. However, I believe the dot-matrix operator expects the target vectors to be rows. So maybe what's missing is another transposition of the dot product to bring the vectors back to column vectors.

effigies replied:
Worth revisiting https://en.wikipedia.org/wiki/Affine_transformation to clear your head. I find it easiest to think of affines as a group; the augmented matrix formulation is identical to the group of all invertible affine transformations. A critical property is that the bottom row is `[0, ..., 0, 1]`, which you lose by transposition.

```Python
In [1]: import numpy as np

In [2]: t1 = np.eye(4)

In [3]: t1[:3, 3] = [1, 2, 3]

In [4]: t2 = 2 * np.eye(4)

In [5]: t2[:, 3] = [4, 5, 6, 1]

In [6]: t1 @ t2
Out[6]: 
array([[2., 0., 0., 5.],
       [0., 2., 0., 7.],
       [0., 0., 2., 9.],
       [0., 0., 0., 1.]])

In [7]: t1 @ t2.T
Out[7]: 
array([[ 6.,  5.,  6.,  1.],
       [ 8., 12., 12.,  2.],
       [12., 15., 20.,  3.],
       [ 4.,  5.,  6.,  1.]])
```

Bearing in mind the group structure helps keep straight that there are only two things you can do: multiply (with the correct product) and invert. Now all you need to keep track of is that the source vector space is on the right and the destination on the left.

oesteban committed:
Update sdcflows/transform.py

Co-authored-by: Chris Markiewicz <markiewicz@stanford.edu>

oesteban committed:
enh: incorporate dense resampling of fieldmap

Requires: #355.

oesteban committed:
sty: add typing annotations and run black

oesteban committed:
fix: miscellaneous problems

oesteban committed:
fix: do not accept multiple files

oesteban committed:
fix: parallelize resampling of 4D images

effigies commented on sdcflows/interfaces/bspline.py:
```diff
@@ -288,116 +288,110 @@ def _run_interface(self, runtime):
+
+    Preconditions:
+
+    * We have a "target EPI" - a BOLD or DWI dataset (or even MPRAGE, same principle),
+      without having gone through HMC or SDC.
+    * We have also the list of HMC matrices that *accounts for* head-motion, so after resampling
+      the dataset through this list of transforms *the head does not move anymore*.
+    * We have estimated the fieldmap's coefficients
+    * We have the "fieldmap-to-data" affine transform that aligns the target dataset (e.g., EPI)
+      and the fieldmap's "magnitude" (phasediff et al.) or "reference" (pepolar, syn).
```
Are we eventually going to add optional transformations from data to final target? That will be the ideal end state where the final resampling to any space includes the fieldmap shift in source space.

oesteban replied:
I think this would be ideal. For that, I believe the best approach would be adding to nitransforms a dense deformation field that works in VOXEL coordinates. There will be some plumbing for composite transforms so that the voxel field is cast to a regular dense field using the (I believe as I write) reference image.

Once we get this right, and I feel it is almost there (just some testing lacking), in my opinion we'll be ready to beef up nitransforms to do that (and perhaps surface too?)

oesteban committed:
ref: cleanup code in preparation of head motion

oesteban committed:
feat: defer head-motion correction into parallel unwarp

This commit defers the handling of head-motion correction transforms to
the separate threads.

It has been tested with transforms having random translations and
everything looks correct.

If passed a constant, zero-fieldmap (coeffs), but a list of realignment
headmotion correction matrices, then the transform should output a
realigned file.

effigies commented on sdcflows/transform.py:
```diff
@@ -72,151 +183,188 @@ def fit(self, spatialimage):
+
+        approx = (
+            approx if affine is not None else False
+        )  # Approximate iff affine is defined
```
```suggestion
        approx &= affine is not None  # Approximate iff affine is defined
```

oesteban committed:
fix: interface internal streamlining toward getting tests to pass

fix: interface internal streamlining toward getting tests to pass

effigies commented on sdcflows/transform.py:
```diff
@@ -72,151 +183,189 @@ def fit(self, spatialimage):
+        # Project the reference's grid onto the fieldmap's
+        target_reference = target_reference.__class__(
+            target_reference.dataobj,
+            affine @ target_affine,
```
When you call `fit(in_data, fmap2data_xfm, approx)`, that gets mapped to `(target_reference, affine, approx)`.

`target_affine` maps target IJK onto target RAS. `affine` maps fmap RAS onto target RAS. The only sensible thing to do here is to map fmap RAS onto target IJK, so we need:

```suggestion
            np.linalg.inv(affine) @ target_affine,
```

Alternately, `fmap2data_xfm` needs to be inverted before passing.

oesteban replied:
I believe this is correct - I agree the nomenclature and the docstring are not clear.

`affine` is the result of calling a registration tool such as ANTs configured for a linear transform with at most 12 degrees of freedom and called with the image carrying `target_affine` as reference and the fieldmap reference as moving.

The result of such a registration framework is an affine (our `affine` here) that maps coordinates in reference (target) RAS on to fieldmap RAS. I'm quite confident this inversion is not needed, but I agree we want to implement some tests here. As I mentioned to @mgxd, I have a good dataset for that :)

For now, what I can do is improving the docstring with something along the lines I have described here. The docstrings need all be revised anyway.

effigies replied:
Sounds good. We need very clear docstrings then. My worry with the second fit called during apply is that this step is being undone in any case, so getting it wrong or right will not have an discernible impact until that bug is fixed.

oesteban replied:
I have added one extra input to the interface, xor'ed with the original one, to pass the reverse transform if that's what you have. Then, inside the interface it is inverted.

effigies commented on sdcflows/transform.py:
```diff
@@ -72,151 +183,189 @@ def fit(self, spatialimage):
-        self.fit(spatialimage)
-        fmap = self.mapped.get_fdata().copy()
+        # Generate warp field (before ensuring positive cosines)
+        self.fit(moving)
```
I think this undoes the `unwarp.fit()` done in bspline.py?

oesteban replied:
It shouldn't - the idea was that it gets cached in the first run, and when the moving has the same characteristics (same matrix and affine) then you can go ahead (even if you actually changed the data because it is matched and you know what you are doing).

This second fit is to make sure we do not apply before checking the moving and fitting the bsplines if necessary.

However, I agree there could be some error in the logic because in my testing it did look like it was undoing the previous fit. I thought it was because I was ensuring positive cosines first. I have it as something I need to double check before merge.

effigies replied:
I think we want to separate fit from validate. Apply should raise an exception on mismatch, not silently drop the important bits.

oesteban replied:
I'm going through the code and I totally agree. Validation should be separate

effigies reviewed (commented):
Couple big issues. Still looking through this.

effigies commented on sdcflows/transform.py:
```diff
@@ -72,151 +183,189 @@ def fit(self, spatialimage):
+
+        # Prepare data
+        data = np.squeeze(np.asanyarray(moving.dataobj))
+        ndim = min(data.ndim, 3)
```
I think you really just want to drop any trailing 1s after the 3rd dimension, correct? On the off chance you do have a 1 in your spatial dimensions, you don't want to invalidate your affine.

```suggestion
        # Squeeze non-spatial dimensions
        newshape = data.shape[:3] + tuple(dim for dim in data.shape[3:] if dim > 1)
        data = np.reshape(moving.dataobj, newshape)
        ndim = min(data.ndim, 3)
```

oesteban replied:
Yes. I had some interest in making this generalizable to 2D but disregarded the idea after realizing it was really hard to infer from the image only and probably there are other decision points where 3D is assumed. Thanks for the improvement.

effigies reviewed (commented):
Okay, I've read through this now. Only one additional comment, and it's more of a nit-pick as series with degenerate spatial dimensions are probably going to break in other ways...

oesteban commented on sdcflows/transform.py:
```diff
@@ -49,21 +142,39 @@ class B0FieldTransform:
+            will be dropped.
+        affine : :obj:`numpy.ndarray`
+            Transform that maps coordinates on the target_reference on to the
+            fieldmap reference (that is, the affine through which the filedmap
```
```suggestion
            fieldmap reference (that is, the affine through which the fieldmap can
```

oesteban committed:
Update sdcflows/transform.py

Co-authored-by: Chris Markiewicz <markiewicz@stanford.edu>

oesteban committed:
Update sdcflows/transform.py

Co-authored-by: Chris Markiewicz <markiewicz@stanford.edu>

oesteban committed:
Update sdcflows/transform.py

oesteban committed:
fix: resolve undefined variable

oesteban committed:
fix/doc: revise docstrings, remove cache validation

CRITICAL change

``sdcflows/interfaces/tests/test_bspline.py::test_bsplines`` is a
parametric test where random coefficients are generated and then re-fit.
The previous resampler was not working.

However, fixing the resampling has made this test a point of concern. We
will need to revisit this test and make sure the bspline fitting is
acceptable.

oesteban committed:
doc: fix documentation and cross-references

oesteban committed:
fix: adapt test with new API of the ApplyCoeffsField interface

oesteban committed:
fix: roll back interface to allow varying pe_dirs/ro_times within a single volume when fmap is shared

oesteban committed:
fix: ensure positive cosines seems necessary before fitting

oesteban committed:
fix: connections in debug version of pipeline

oesteban committed:
fix: test on more frequentist terms

mgxd commented on sdcflows/interfaces/bspline.py:
```diff
@@ -288,116 +284,112 @@ def _run_interface(self, runtime):
+        True,
+        usedefault=True,
+        desc=(
+            "reconstruct the fieldmap on it's original grid and then interpolate on the "
```
```suggestion
            "reconstruct the fieldmap on its original grid and then interpolate on the "
```

mgxd commented on sdcflows/interfaces/bspline.py:
```diff
@@ -288,116 +284,112 @@ def _run_interface(self, runtime):
+            num_threads=(
+                None if not isdefined(self.inputs.num_threads) else self.inputs.num_threads
+            ),
+        ).to_filename(self._results["out_corrected"])
```
to keep a consistent syntax
```suggestion
        unwarp.apply(
            self.inputs.in_data,
            self.inputs.pe_dir,
            self.inputs.ro_time,
            xfms=self.inputs.in_xfms if isdefined(self.inputs.in_xfms) else None,
            fmap2data_xfm=(
                self.inputs.fmap2data_xfm if isdefined(self.inputs.fmap2data_xfm) else None
            ),
            approx=self.inputs.approx,
            num_threads=(
                self.inputs.num_threads if isdefined(self.inputs.num_threads) else None
            ),
        ).to_filename(self._results["out_corrected"])
```

mgxd commented on sdcflows/transform.py:
```diff
@@ -32,15 +59,166 @@
+    data: np.ndarray,
+    coordinates: np.ndarray,
+    pe_info: Tuple[int, float],
+    hmc_xfm: np.ndarray,
```
since the in/ex-clusion of this is important
```suggestion
    hmc_xfm: Optional[np.ndarray],
```

mgxd commented on sdcflows/transform.py:
```diff
@@ -20,8 +20,35 @@
+from functools import partial
+import asyncio
 from pathlib import Path
+from typing import Callable, List, Sequence, Tuple, Union
```
see below
```suggestion
from typing import Callable, List, Optional, Sequence, Tuple, Union
```

mgxd commented on sdcflows/transform.py:
```diff
@@ -72,151 +274,243 @@ def fit(self, spatialimage):
 
         # Resample
-        resampled = ndi.map_coordinates(
+        resampled = asyncio.run(unwarp_parallel(
```
I'm curious to see how performance is affected by switching this to work within an async framework - just noting that this will limit the use of this class/function in cases of an already running event loop (without hacks like nest_asyncio)

oesteban replied:
I'd say not a lot better than a thread executor, but much better than a processor executor or joblib

oesteban replied:
Here the main bottleneck is memory.

mgxd reviewed (commented):
A few passing comments

oesteban committed:
Apply suggestions from code review

Co-authored-by: Mathias Goncalves <mathiasg@stanford.edu>

oesteban committed:
fix: add connection to feed fmap2data transform

oesteban committed:
maint: add test of fit-pepolar/apply workflows

oesteban committed:
fix: revise type of input to bspline interface

oesteban committed:
fix: error in test because of connection removed

oesteban committed:
maint: add unittest on B0FieldTransform with hcph data

oesteban committed:
fix/feat: add GRE fieldmaps of hcph to integration tests (fix applycoeffs interface)

oesteban committed:
sty: run black on changed files

oesteban committed:
fix(wip): forward the fmap2epi transform

effigies commented on sdcflows/transform.py:
```diff
@@ -49,174 +227,291 @@ class B0FieldTransform:
             ``False`` if cache was valid and will be reused.
 
         """
+        import pdb; pdb.set_trace()
```
This is causing CircleCI to hang indefinitely. We should probably remove `-it` from `docker run` commands to avoid this.

```suggestion
```

effigies committed:
Remove pdb trace

oesteban committed:
fix: apply cannot be done on np.ndarray

oesteban committed:
maint: update nitransforms pin

effigies commented on sdcflows/transform.py:
```diff
@@ -32,15 +59,166 @@
+    prefilter: bool = True,
+    output_dtype: Union[str, np.dtype] = None,
+    max_concurrent: int = min(os.cpu_count(), 12),
+) -> List[np.ndarray]:
```
```suggestion
) -> np.ndarray:
```

oesteban committed:
enh: do not approx if coefficients and target grid are aligned

oesteban committed:
maint: add PA/AP tests with one more session of data

oesteban committed:
fix: revise registration workflow

Since the affine between the target EPI and the fieldmap reference is
now applied "on-the-fly", the coefficients SHOULD NOT be reoriented
into the target space.

This also revises the test on the correction workflow, which also ran
registration (stacking potential problems by exercising several units)
so now the alignment matrix has been calculated offline and we can run
the test with and without it.

This commit requires the derivatives of the HCP dataset be up-to-date.

oesteban committed:
enh: add parameterized test on the transform

oesteban committed:
fix+enh: wrong transform creating NIfTI object + move to transform points convention

The wrong affine was assigned to the NIfTI object carrying the resampled
fieldmap, which did not fail if no transform between fieldmap and EPI
was given.

This PR also changes the naming of variables from BIDS' ``mode-image``
convention to ``mode-points`` convention, as the latter is a definition
closer to what we are actually doing (transforming points).

mgxd commented on sdcflows/workflows/apply/correction.py:
```diff
@@ -148,21 +151,18 @@ def init_unwarp_wf(*, free_mem=None, omp_nthreads=1, debug=False, name="unwarp_w
-                                    ("out_warp", "fieldwarp_ref")]),
+        (resample, outputnode, [("out_field", "fieldmap")]),
+        # (resample_ref, outputnode, [("out_field", "fieldmap_ref")]),
+        # TODO - take reference from brainextraction workflow
```
still todo?

oesteban committed:
enh: disable head-motion transforms consideration

effigies commented on sdcflows/interfaces/bspline.py:
```diff
@@ -288,116 +292,137 @@ def _run_interface(self, runtime):
+        desc="the transform by which the fieldmap can be resampled on the target EPI's grid.",
+        xor="fmap2data_xfm",
+    )
+    in_xfms = traits.List(
+        traits.List(traits.List(traits.Float)),
+        desc="list of head-motion correction matrices",
     )
```
Note to self: Figure out why this should not be files.

mgxd commented on sdcflows/interfaces/bspline.py:
```diff
@@ -288,116 +292,137 @@ def _run_interface(self, runtime):
     )
-    in_xfms = InputMultiObject(
-        File(exists=True), desc="list of head-motion correction matrices"
+    fmap2data_xfm = InputMultiObject(
+        File(exists=True),
+        desc="the transform by which the target EPI can be resampled on the fieldmap's grid.",
+        xor="data2fmap_xfm",
+    )
+    data2fmap_xfm = InputMultiObject(
+        File(exists=True),
+        desc="the transform by which the fieldmap can be resampled on the target EPI's grid.",
+        xor="fmap2data_xfm",
+    )
```
Seems confusing making these MultiObjects, since the first is always being used and the rest discarded - and we should specify that these need to be in ITK format.

oesteban replied:
This is for compatiblity with the nipype interface. We probably want not to drop linear transforms after the first, although it feels like a rabbit hole.

So, rather than operating here, I think we want to open an issue to think about how a list here should be managed later.

oesteban replied:
To follow up in #386

mgxd commented on sdcflows/interfaces/bspline.py:
```diff
@@ -431,7 +458,8 @@ def _run_interface(self, runtime):
-                    self.inputs.fmap_target if isdefined(self.inputs.fmap_target)
+                    self.inputs.fmap_target
+                    if isdefined(self.inputs.fmap_target)
                     else None
```
```suggestion
                    self.inputs.fmap_target or None
```

oesteban replied:
Ditto

mgxd commented on sdcflows/interfaces/bspline.py:
```diff
@@ -288,116 +292,137 @@ def _run_interface(self, runtime):
+            approx=self.inputs.approx,
+            num_threads=(
+                self.inputs.num_threads if isdefined(self.inputs.num_threads) else None
+            ),
```
less verbose
```suggestion
            xfms=self.inputs.in_xfms or None,
            xfm_data2fmap=data2fmap_xfm,
            approx=self.inputs.approx,
            num_threads=self.inputs.num_threads or None,
```

oesteban replied:
nipype's `Undefined` doesn't evaluate to False, as far as I know.

mgxd replied:
```python
In [1]: from nipype.interfaces.base import Undefined

In [2]: Undefined or True
Out[2]: True
```

effigies replied:
```
In [3]: from nipype.interfaces.base import Undefined

In [4]: bool(Undefined)
Out[4]: False
```

mgxd commented on sdcflows/interfaces/tests/test_bspline.py:
```diff
@@ -80,8 +80,10 @@ def test_bsplines(tmp_path, testnum):
-    # Absolute error of the interpolated field is always below 5 Hz
-    assert np.all(np.abs(nb.load(test2.outputs.out_error).get_fdata()) < 5)
+    # Absolute error of the interpolated field
+    # TODO - this is probably too high. We need to revisit these tests.
```
is this still TODO?

oesteban replied:
yes

oesteban replied:
see #369

mgxd commented on sdcflows/transform.py:
```diff
@@ -20,8 +20,35 @@
+
+
+"""
+import os
```
We should take advantage of the free backport to get cleaner annotations, allowing us to get the behavior in the comments below
```suggestion
"""
from __future__ import annotations

import os
```

effigies replied:
Note that while we use `from __future__ import annotations`, we will not be able to use type annotations with `attrs` in this file. We do not currently, so this is fine.

mgxd commented on sdcflows/transform.py:
```diff
@@ -32,15 +59,166 @@
+    pe_info: Tuple[int, float],
+    hmc_xfm: Optional[np.ndarray],
+    fmap_hz: np.ndarray,
+    output_dtype: Union[type, np.dtype] = None,
```
is `output_dtype` meant to be `str`?
```suggestion
    hmc_xfm: np.ndarray | None,
    fmap_hz: np.ndarray,
    output_dtype: str | np.dtype | None = None,
```

mgxd commented on sdcflows/transform.py:
```diff
@@ -32,15 +59,166 @@
+    mode: str = "constant",
+    cval: float = 0.0,
+    prefilter: bool = True,
+    output_dtype: Union[str, np.dtype] = None,
```
```suggestion
    output_dtype: str | np.dtype | None = None,
```

mgxd commented on sdcflows/transform.py:
```diff
@@ -49,21 +227,45 @@ class B0FieldTransform:
+    def fit(
+        self,
+        target_reference: nb.spatialimages.SpatialImage,
+        xfm_data2fmap: np.ndarray = None,
```
```suggestion
        xfm_data2fmap: np.ndarray | None = None,
```

mgxd commented on sdcflows/transform.py:
```diff
@@ -72,151 +274,271 @@ def fit(self, spatialimage):
+            )
 
         # Generate tensor-product B-Spline weights
         for level in listify(self.coeffs):
```
`listify` is potentially called 3 times on `self.coeffs` - perhaps we save the cycles and just set to a variable?

mgxd commented on sdcflows/transform.py:
```diff
@@ -72,151 +274,271 @@ def fit(self, spatialimage):
+        #     if np.all(newshape == self.mapped.shape) and np.allclose(
+        #         target_affine, self.mapped.affine
+        #     ):
+        #         return False
```
can this be removed since `self.mapped` is now present?

oesteban replied:
```suggestion
```

mgxd commented on sdcflows/transform.py:
```diff
@@ -72,151 +274,271 @@ def fit(self, spatialimage):
+        mode: str = "constant",
+        cval: float = 0.0,
+        prefilter: bool = True,
+        output_dtype: Union[str, np.dtype] = None,
```
```suggestion
        pe_dir: str | Sequence[str],
        ro_time: float | Sequence[float],
        xfms: Sequence[np.ndarray] | None = None,
        xfm_data2fmap: np.ndarray | None = None,
        approx: bool = True,
        order: int = 3,
        mode: str = "constant",
        cval: float = 0.0,
        prefilter: bool = True,
        output_dtype: str, np.dtype | None = None,
```

mgxd commented on sdcflows/workflows/apply/registration.py:
```diff
@@ -145,27 +145,7 @@ def init_coeff2epi_wf(
-        ])
-        # fmt: on
+    if write_coeff:
+        warn("SDCFlows does not tinker with the coefficients file anymore")
```
```suggestion
        warn(
            "SDCFlows does not tinker with the coefficients file anymore, " 
            "the `write_coeff` parameter will be removed in a future release."
        )
```

mgxd reviewed (commented):
gave another quick pass

effigies commented on sdcflows/transform.py:
```diff
@@ -72,151 +274,271 @@ def fit(self, spatialimage):
+            target_reference.dataobj,
+            xfm_data2fmap @ target_affine,
+            target_header,
+        )
```
This feels hard to follow. We use `target_affine` and `target_header` late on to get back to the original reference. What if we create a new image instead:

```suggestion
        # Project the reference's grid onto the fieldmap's
        projected_reference = target_reference.__class__(
            target_reference.dataobj,
            xfm_data2fmap @ target_reference.affine,
            target_reference.header,
        )
```

This would make it easier to keep track of which "target reference" is intended at all points.

effigies commented on sdcflows/transform.py:
```diff
@@ -72,151 +274,271 @@ def fit(self, spatialimage):
-        # Initialize xform (or set identity)
-        xfm = self.xfm if self.xfm is not None else nt.linear.Affine()
+        # Make sure the data array has all cosines positive (i.e., no axes are flipped)
+        target_reference, _ = ensure_positive_cosines(target_reference)
```
```suggestion
        projected_reference, _ = ensure_positive_cosines(projected_reference)
```

effigies commented on sdcflows/transform.py:
```diff
@@ -72,151 +274,271 @@ def fit(self, spatialimage):
+        # Make sure the data array has all cosines positive (i.e., no axes are flipped)
+        target_reference, _ = ensure_positive_cosines(target_reference)
+
+        # Approximate only iff the coordinate systems are not aligned
```
Nit-pick

```suggestion
        # Approximate only if the coordinate systems are not aligned
```

effigies commented on sdcflows/transform.py:
```diff
@@ -72,151 +274,271 @@ def fit(self, spatialimage):
         coeffs = []
+        _tmp_reference = None
+
+        hdr = target_reference.header.copy()
```
We do not need to declare `_tmp_reference`. It's only used in one `if block`.

This `header.copy()` is too far from its modification. It should also be identical to the original `target_reference.header`, except the affine, so it's okay not to preserve it before modifying `projected_reference` here.

```suggestion
```

effigies commented on sdcflows/transform.py:
```diff
@@ -72,151 +274,271 @@ def fit(self, spatialimage):
+
+            # Generate a sampling reference on the fieldmap's space that fully covers
+            # the target_reference's grid.
+            target_reference = deoblique_and_zooms(
```
```suggestion
            projected_reference = deoblique_and_zooms(
```

effigies commented on sdcflows/transform.py:
```diff
@@ -72,151 +274,271 @@ def fit(self, spatialimage):
-        vsm = (np.squeeze(np.hstack(coeffs).T) @ sparse_vstack(weights)).reshape(
-            vsm.shape
+        # Reconstruct the fieldmap (in Hz) from coefficients
+        fmap = np.zeros(target_reference.shape[:3], dtype="float32")
```
```suggestion
        fmap = np.zeros(projected_reference.shape[:3], dtype="float32")
```

effigies commented on sdcflows/transform.py:
```diff
@@ -72,151 +274,271 @@ def fit(self, spatialimage):
+        hdr["cal_min"] = -hdr["cal_max"]
+
+        # Cache
+        self.mapped = nb.Nifti1Image(fmap, target_reference.affine, hdr)
```
```suggestion
        self.mapped = nb.Nifti1Image(fmap, projected_reference.affine, hdr)
```

effigies commented on sdcflows/transform.py:
```diff
@@ -72,151 +274,271 @@ def fit(self, spatialimage):
-        hdr = spatialimage.header.copy()
-        hdr.set_intent("estimate", name="Voxel shift")
+        # Generate a NIfTI object
+        hdr.set_intent("estimate", name="fieldmap Hz")
```
```suggestion
        hdr = target_reference.header.copy()
        hdr.set_intent("estimate", name="fieldmap Hz")
```

effigies commented on sdcflows/transform.py:
```diff
@@ -72,151 +274,271 @@ def fit(self, spatialimage):
+            _tmp_reference = nb.Nifti1Image(
+                np.zeros(_tmp_shape, dtype=target_header.get_data_dtype()),
+                target_affine,
+                target_header,
```
Does the dtype of the data array matter?

```suggestion
                np.zeros(target_reference.shape[:3], dtype=target_reference.get_data_dtype()),
                target_reference.affine,
                target_reference.header,
```

effigies commented on sdcflows/transform.py:
```diff
@@ -72,151 +274,271 @@ def fit(self, spatialimage):
+            # Generate warp field (before ensuring positive cosines)
+            self.fit(moving, xfm_data2fmap=xfm_data2fmap, approx=approx)
+
+        # Squeeze n33on-spatial dimensions
```
```suggestion
        # Squeeze non-spatial dimensions
```

effigies commented on sdcflows/transform.py:
```diff
@@ -72,151 +274,271 @@ def fit(self, spatialimage):
+
+        # Squeeze n33on-spatial dimensions
+        newshape = moving.shape[:3] + tuple(dim for dim in moving.shape[3:] if dim > 1)
+        data = np.reshape(moving.dataobj, newshape)
```
If it saves anything to avoid loading data into RAM this early, you can reshape array proxies without loading them:

```suggestion
        data = nb.arrayproxy.reshape_dataobj(moving.dataobj, newshape)
```

Feel free to close if it doesn't make a difference.

effigies commented on sdcflows/workflows/apply/correction.py:
```diff
@@ -135,8 +140,6 @@ def init_unwarp_wf(*, free_mem=None, omp_nthreads=1, debug=False, name="unwarp_w
         name="resample",
     )
 
-    resample_ref = pe.Node(ApplyCoeffsField(), mem_gb=mem_per_thread, name="resample_ref")
-
```
We added this because it is useful for reporting, and I intend to use it for the fit/apply (see #385). Is the alternative here to pass the reference as `distorted` with no `hmc_xforms`?

oesteban replied:
Wouldn't that be more explicit? The workflow is the same for both boldref and bold, except for `hmc_xforms`, as you mention.

effigies replied:
Yeah, that's fine with me. Will close #385.

effigies commented on sdcflows/interfaces/bspline.py:
```diff
@@ -288,116 +292,137 @@ def _run_interface(self, runtime):
-        if len(out_fields) == 1:
-            self._results["out_field"] = out_fields.pop()
+        # HMC matrices are only necessary when reslicing the data (i.e., apply())
+        # Check the length of in_xfms matches that of in_data
```
This comment doesn't seem to apply.

oesteban replied:
```suggestion
```

effigies commented on sdcflows/workflows/tests/test_integration.py:
```diff
@@ -0,0 +1,220 @@
+
+    session = "15" if pe0 == "LR" else "14"
+
+    pe1 = f"{pe0[::-1]}"
```
```suggestion
    pe1 = pe0[::-1]
```

effigies commented on sdcflows/workflows/tests/test_integration.py:
```diff
@@ -0,0 +1,220 @@
+    wf = pe.Workflow(name=f"hcph_{mode}_{pe0}")
+
+    # Execute in temporary directory
+    wf.base_dir = f"{tmpdir}" if not workdir else str(workdir)
```
Weirdly convoluted.

```suggestion
    wf.base_dir = str(workdir or tmpdir)
```

effigies reviewed (commented):
Okay. Overall looks good. Most suggestions are for maintainability, not bugs caught.

oesteban commented on sdcflows/transform.py:
```diff
@@ -72,151 +274,271 @@ def fit(self, spatialimage):
+        # TODO Separate cache validation from here. With the resampling, this
+        # code will always  determine the cache must be recalculated.
+        # if self.mapped is not None:
+        #     newshape = target_reference.shape
```
```suggestion
```

oesteban committed:
Apply suggestions from code review

Co-authored-by: Mathias Goncalves <mathiasg@stanford.edu>
Co-authored-by: Chris Markiewicz <markiewicz@stanford.edu>

oesteban committed:
fix+sty: argument specification + run black

effigies commented on setup.cfg:
```diff
@@ -36,7 +36,7 @@ install_requires =
     nitransforms >= 23.0.1
     numpy >= 1.21.0
-    pybids >= 0.15.1
+    pybids >= 0.15.1, <= 0.16.1
```
Maximums are painful. Can I suggest a `!=` to allow for bug fixes?

```suggestion
    pybids >= 0.15.1, != 0.16.2
```

oesteban committed:
Apply suggestions from code review

Co-authored-by: Mathias Goncalves <mathiasg@stanford.edu>

oesteban committed:
enh: address comments from code review

Co-authored-by: Mathias Goncalves <mathiasg@stanford.edu>
